### PR TITLE
[autoparallel] add numerical test for node strategies

### PR DESCRIPTION
**Pull request by YuliangLiu0306**
### What does this PR do

This PR provides some apis to do numerical tests for node strategies. The function called numerical_test_for_node_strategy will check the **output of forward result, the gradient of the input activation and gradient of parameters if exist** with the original result of non-distributed version.

Note: only convolution handler tests updated with the above apis as an example, other handlers unit test will be updated in next PR.

### CODE REVIEW

File: colossalai/auto_parallel/passes/runtime_apply_pass.py [hunk 1]
```diff
@@ -81,18 +80,22 @@ def _shape_consistency_apply(gm: torch.fx.GraphModule):
         if not hasattr(node, 'best_strategy') or node.op == 'output':
             continue
 
-        for user_node in node.strategies_vector.successor_nodes:
-            user_node_index = user_node.strategies_vector.predecessor_nodes.index(node)
+        for user_node_index, user_node in enumerate(node.strategies_vector.successor_nodes):
+            # user_node_index = user_node.strategies_vector.predecessor_nodes.index(node)
```
Comment (FrankLeeeee) [inline colossalai/auto_parallel/passes/runtime_apply_pass.py:84]: remove this line?

File: colossalai/auto_parallel/tensor_shard/sharding_strategy.py [hunk 2]
```diff
@@ -115,6 +115,7 @@ class CommAction:
     comm_spec: CommSpec = None
     comm_type: CommType = None
     arg_index: int = -1
+    key_for_kwarg: any = None
```
Comment (FrankLeeeee) [inline colossalai/auto_parallel/tensor_shard/sharding_strategy.py:118]: what is this used for?

File: colossalai/auto_parallel/passes/runtime_apply_pass.py [hunk 3]
```diff
@@ -112,18 +115,29 @@ def _comm_spec_apply(gm: torch.fx.GraphModule):
...
-                node.args = new_args
+
+                if comm_action.key_for_kwarg is not None:
+                    new_kwargs = dict(node.kwargs)
+                    new_kwargs[comm_action.key_for_kwarg] = comm_spec_apply_node
+                    node.kwargs = new_kwargs
+                else:
+                    new_args = list(node.args)
+                    new_args[comm_action.arg_index] = comm_spec_apply_node
+                    node.args = new_args
```
Comment (FrankLeeeee) [inline colossalai/auto_parallel/passes/runtime_apply_pass.py:139]: It is kind of difficult to understand what is going on here. Can you write some comments (not just here, but other places as well)?

File: tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
+import copy
+from re import T
```
Comment (FrankLeeeee) [inline tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py:2]: I think this import is not used.

File: colossalai/auto_parallel/tensor_shard/sharding_strategy.py (hunk 2 above)
Comment (YuliangLiu0306) [reply colossalai/auto_parallel/tensor_shard/sharding_strategy.py:118]: If a node is kwarg of other node, not a positional argument, we could use this to locate it in runtime pass.

File: colossalai/auto_parallel/passes/runtime_apply_pass.py (hunk 1 above)
Comment (YuliangLiu0306) [reply colossalai/auto_parallel/passes/runtime_apply_pass.py:84]: Sure

File: tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py (hunk 4 above)
Comment (YuliangLiu0306) [reply tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py:2]: I will remove these imports.

File: tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            assert_close((global_grad - grad_to_compare).sum(), zero_tensor)
+
+
+def to_global(distributed_tensor: torch.Tensor, sharding_spec: ShardingSpec):
+    shape_consistency_manager = ShapeConsistencyManager()
+    global_sharding_spec = ShardingSpec(sharding_spec.device_mesh, sharding_spec.entire_shape, {})
+    with torch.no_grad():
+        global_tensor = shape_consistency_manager.apply_for_autoparallel_runtime(distributed_tensor, sharding_spec,
+                                                                                 global_sharding_spec)
+    return global_tensor
```
Comment (FrankLeeeee) [inline tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py:137]: This method should be put in the `shariding_spec.py` so that it can be possibly used by other files in the future.

File: colossalai/auto_parallel/passes/runtime_apply_pass.py (hunk 3 above)
Comment (YuliangLiu0306) [reply colossalai/auto_parallel/passes/runtime_apply_pass.py:139]: Sure

File: tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py (hunk 5 above)
Comment (YuliangLiu0306) [reply tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py:137]: Okay, I will move it into sharding_spec.py.

File: tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        wrapper(kwarg_to_compare, name)
+        kwargs_to_compare[name] = kwarg_to_compare
+
+    return model_to_compare, args_to_compare, kwargs_to_compare
+
+
+def numerical_test_for_node_strategy(model: torch.nn.Module,
+                                     device_mesh: DeviceMesh,
+                                     node_index: int,
+                                     strategy_number: int,
```
Comment (FrankLeeeee) [inline tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py:59]: can this number be set to 1 first and be automatically updated after calling `build_strategies_and_cost`?

Comment (FrankLeeeee) [reply tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py:59]: so that we don't have to manually count how many strategies to test.

Comment (YuliangLiu0306) [reply tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py:59]: Yes, but I think strategy number should be a consistent for unit tests.

Comment (FrankLeeeee) [reply tests/test_auto_parallel/test_tensor_shard/test_node_handler/utils.py:59]: Okay, one possible omission is that the tester gives a wrong number (e.g. lower than the actual number), then the test is not complete.

**Review by FrankLeeeee — APPROVED** [review]
